merkle/src/store.rs

```diff
@@ -221,6 +349,49 @@ impl<E: Element> Store<E> for DiskStore<E> {
         self.len
     }
 
+    // Specifically, this method truncates an existing DiskStore and
+    // formats the data in such a way that is compatible with future
+    // access using LevelCacheStore::new_from_disk.
+    fn compact(&mut self, config: StoreConfig) -> bool {
+        // Determine how many leafs there are (in bytes).
+        let data_width = (self.len / 2 + 1) * self.elem_len;
+
+        // Calculate how large the cache should be (based on the
+        // config.levels param) and load that much data from the end
+        // of the store into RAM temporarily.
```

Comment:
instead of loading this into ram, can't you do a copy inside the file

(doesn't compile, just to illustrate what I mean)

```rust
let file1 = File::open_read(path)?;
let file2 = File::open_write(path)?;

file1.seek(SeekFrom::Start(cache_offset))?;
file2.seek(SeekFrom::Start(start_of_tree))?;

io::copy(file1.take(cache_size), file2)?;
```

Reply:
technically this still goes through ram, but should be more efficient, not require a full allocation

Reply:
I don't quite understand this (or maybe don't know the right rust calls), but from my understanding, the io::copy needs something that already has data read in (i.e. something allocated that read it) in order to write it into the same file.

The `file1.take(cache_size)` seems nice, but I think it would have to be something like `file1.read_to_end(some_allocated_buffer).take(cache_size)` instead, which defeats the purpose.

I like the idea, just not seeing how to do it.  I'm perhaps missing the obvious here.

Reply:
> the io::copy needs something that already has data read in 

Luckily it does not. I wrote up an example that does this, so you can see what I meant: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6557e90c0519b00cfce0f2c1c1460812

Reply:
Very cool, thanks!